### PR TITLE
[FEAT] Keystone integration

### PR DESCRIPTION
**Pull request by gantunesr**
**Description**

Keystone integration

**References**

- [Keystone General Guide](https://support.keyst.one/)
- [Keystone QR Code Protocol](https://github.com/KeystoneHQ/Keystone-developer-hub/blob/main/research/ethereum-qr-data-protocol.md)
-  [Keystone - ur-registry](https://github.com/KeystoneHQ/ur-registry)
- [Keystone - airgaped-base](https://github.com/KeystoneHQ/keystone-airgaped-base)
  - [Keystone - metamask-airgaped-keyring](https://github.com/KeystoneHQ/keystone-airgaped-base/tree/master/packages/metamask-airgapped-keyring)

### CODE REVIEW

File: app/components/UI/AccountOverview/index.js [hunk 1]
```diff
@@ -201,6 +222,8 @@ class AccountOverview extends PureComponent {
 
 	input = React.createRef();
 
+	KeyringController = Engine.context.KeyringController;
```
Comment (gantunesr) [inline app/components/UI/AccountOverview/index.js:225]: Remove

File: app/components/UI/ApproveTransactionReview/index.js [hunk 2]
```diff
@@ -353,13 +363,14 @@ class ApproveTransactionReview extends PureComponent {
 
 	getAnalyticsParams = () => {
 		try {
-			const { activeTabUrl, transaction, onSetAnalyticsParams } = this.props;
+			const { activeTabUrl, transaction, onSetAnalyticsParams, selectedAddress } = this.props;
 			const { tokenSymbol, originalApproveAmount, encodedAmount } = this.state;
 			const { NetworkController } = Engine.context;
 			const { chainId, type } = NetworkController?.state?.provider || {};
 			const isDapp = !Object.values(AppConstants.DEEPLINKS).includes(transaction?.origin);
 			const unlimited = encodedAmount === 'ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff';
```
Comment (gantunesr) [inline app/components/UI/ApproveTransactionReview/index.js:371]: Move to constant

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	onScanSuccess: (ur: UR) => void;
+	onScanError: (error: string) => void;
+	hideModal: () => void;
+}
+
+export const SupportedURType = {
+	CRYPTO_HDKEY: 'crypto-hdkey',
+	CRYPTO_ACCOUNT: 'crypto-account',
+	ETH_SIGNATURE: 'eth-signature',
+};
```
Comment (gantunesr) [inline app/components/UI/QRHardware/AnimatedQRScanner.tsx:87]: Move this constants to another file

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+const AnimatedQRScannerModal = (props: AnimatedQRScannerProps) => {
+	const { visible, onScanError, purpose, onScanSuccess, hideModal } = props;
+	const [urDecoder, setURDecoder] = useState(new URRegistryDecoder());
+	const [progress, setProgress] = useState(0);
+
+	let expectedURTypes: string[];
+	if (purpose === 'sync') {
```
Comment (gantunesr) [inline app/components/UI/QRHardware/AnimatedQRScanner.tsx:95]: Move to constant

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+		),
+		[purpose]
+	);
+
+	const getAnalyticParams = useCallback(
+		() => ({
+			purpose,
+		}),
+		[purpose]
+	);
```
Comment (gantunesr) [inline app/components/UI/QRHardware/AnimatedQRScanner.tsx:125]: Remove hook

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,18071 @@
+{
```
Comment (gantunesr) [inline package-lock.json:1]: Remove file

File: ios/Podfile.lock [hunk 7]
```diff
@@ -841,4 +841,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: 5759da19ce141e5f8aca3a4f56ac232baa34e957
 
-COCOAPODS: 1.11.2
+COCOAPODS: 1.10.1
```
Comment (gantunesr) [inline ios/Podfile.lock:844]: revert

**Review by CryptoGodfatherVA4 — APPROVED** [review]

**Review by Cal-L — COMMENTED** [review]
Left comments :). The comments below are duplicate comments. Refer to the comments above.

File: app/components/UI/TransactionElement/index.js [hunk 8]
```diff
@@ -269,12 +273,18 @@ class TransactionElement extends PureComponent {
...
 						)}
 					</ListItem.Content>
-					{!!renderTxActions && (
+					{!!renderNormalActions && (
 						<ListItem.Actions>
 							{this.renderSpeedUpButton()}
 							{this.renderCancelButton()}
 						</ListItem.Actions>
 					)}
+					{!!renderUnsignedQRActions && (
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/TransactionElement/index.js]: [f6fe256](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/f6fe2561e74c1f33bb185ffc06ea004fbc694763)

File: app/components/Views/ConnectQRHardware/index.tsx [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+				};
+			}),
+		[accounts, checkedAccounts, existingAccounts, trackedAccounts]
+	);
+
+	const onUnlock = useCallback(async () => {
+		resetError();
+		setBlockingModalVisible(true);
+		for (const account of checkedAccounts) {
+			await KeyringController.unlockQRHardwareWalletAccount(account);
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/index.tsx]: [bd33425](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/bd33425612cde320058ea7ad5cd5029eaec52a47)

File: app/components/UI/Navbar/index.js [hunk 10]
```diff
@@ -453,7 +453,7 @@ export function getSendFlowTitle(title, navigation, route, themeColors) {
 	};
 	const leftAction = () => navigation.pop();
 
-	const canGoBack = title !== 'send.send_to' && !route?.params?.isPaymentRequest;
+	const canGoBack = title !== 'send.send_to' && !route?.params?.isPaymentRequest && title !== 'send.sign';
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/Navbar/index.js]: [bf74a60](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/bf74a606855a7aab66618d00e62b382103b53ae3)

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+							style={styles.checkBox}
+							disabled={item.exist}
+							value={item.checked}
+							onValueChange={() => toggleAccount(item.index)}
+							boxType={'square'}
+							tintColors={{ true: colors.background.default, false: colors.background.default }}
+							testID={'skip-backup-check'}
+						/>
+						<Text style={styles.number}>{item.index}</Text>
+						<Text style={styles.address}>{util.clipAddress(item.address, 4, 4)}</Text>
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx]: [482a80e](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/482a80e253d764afa38ff7c581a6b0c2898d6b27)

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/util.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+const clipAddress = (address: string, showHead: number, showTail: number) => {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/util.ts]: https://github.com/MetaMask/metamask-mobile/pull/3915/commits/d7b92b65a4d1db0b2504adbecb043ba2699b6710

File: app/components/Nav/Main/MainNavigator.js [hunk 13]
```diff
@@ -321,6 +322,16 @@ const SetPasswordFlow = () => (
 	</Stack.Navigator>
 );
 
+const ConnectQRHardwareFlow = () => (
+	<Stack.Navigator
+		screenOptions={{
+			headerShown: false,
+		}}
+	>
+		<Stack.Screen name="ConnectQRHardware" component={ConnectQRHardware} />
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Nav/Main/MainNavigator.js:331]: Noted for next release

File: app/components/Views/TransactionsView/index.js [hunk 14]
```diff
@@ -46,8 +46,9 @@ const TransactionsView = ({
 		const confirmedTxs = [];
 		const submittedNonces = [];
 
-		const allTransactionsSorted = sortTransactions(transactions);
-
+		const allTransactionsSorted = sortTransactions(transactions).filter(
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/TransactionsView/index.js:49]: Noted for next release. Solution will be moved to controllers

File: app/components/Views/Asset/index.js [hunk 15]
```diff
@@ -205,7 +205,9 @@ class Asset extends PureComponent {
 
 		const { chainId, transactions } = this.props;
 		if (transactions.length) {
-			const sortedTransactions = sortTransactions(transactions);
+			const sortedTransactions = sortTransactions(transactions).filter(
+				(tx, index, self) => self.findIndex((_tx) => _tx.id === tx.id) === index
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/Asset/index.js:209]: Noted for next release. Solution will be moved to controllers

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx [hunk 16]
```diff
@@ -0,0 +1,188 @@
...
+		},
+		backgroundGrey: {
+			backgroundColor: colors.primary.disabled,
+		},
+		textBlue: {
+			fontSize: 14,
+			color: colors.primary.default,
+			...fontStyles.normal,
+		},
+		textWhite: {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx]: [482a80e](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/482a80e253d764afa38ff7c581a6b0c2898d6b27)

File: app/components/Views/ConnectQRHardware/Instruction/index.tsx [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+					<Text style={styles.text}>{strings('connect_qr_hardware.description3')}</Text>
+					<Text style={[styles.text, styles.link]} onPress={navigateToTutorial}>
+						{strings('connect_qr_hardware.description4')}
+					</Text>
+					<Text style={styles.text}>{strings('connect_qr_hardware.description5')}</Text>
+					<Text style={styles.text}>{strings('connect_qr_hardware.description6')}</Text>
+				</View>
+				<Image style={styles.image} source={connectQRHardwareImg} resizeMode={'contain'} />
+			</ScrollView>
+			<TouchableOpacity onPress={onConnect} style={styles.button}>
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/Instruction/index.tsx]: [0ccf01a](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/0ccf01a18aa364614c0298734ea13e0ba0213ca1)

File: app/components/Views/ConnectQRHardware/Instruction/index.tsx [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+				url: 'https://keyst.one/mmmvideo',
+				title: strings('connect_qr_hardware.description2'),
+			},
+		});
+	};
+	const navigateToTutorial = () => {
+		navigation.navigate('Webview', {
+			screen: 'SimpleWebview',
+			params: {
+				url: 'https://keyst.one/mmm',
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/Instruction/index.tsx]: [43fbdfc](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/43fbdfc594ed946d96e179f003997982893e36e2)

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx [hunk 19]
```diff
@@ -0,0 +1,188 @@
...
+
+const SelectQRAccounts = (props: ISelectQRAccountsProps) => {
+	const { accounts, prevPage, nextPage, toggleAccount, onForget, onUnlock, canUnlock } = props;
+	const { colors } = useAppThemeFromContext() || mockTheme;
+	const styles = createStyle(colors);
+	const navigation = useNavigation();
+	const chainId = useSelector((state: any) => state.engine.backgroundState.NetworkController.provider.chainId);
+
+	const toEtherscan = (address: string) => {
+		const accountLink = getEtherscanAddressUrl(getNetworkTypeByChainId(chainId), address);
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx]: [4665b74](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/4665b74b580f55cd48ba99559617a8a579813ea7)

File: app/util/networks/index.js [hunk 20]
```diff
@@ -102,6 +103,18 @@ export function getNetworkTypeById(id) {
 	throw new Error(`${NETWORK_ERROR_UNKNOWN_NETWORK_ID} ${id}`);
 }
 
+export function getNetworkTypeByChainId(chainId) {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/util/networks/index.js]: [4665b74](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/4665b74b580f55cd48ba99559617a8a579813ea7)

File: app/components/Views/ConnectQRHardware/index.tsx [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+	return (
+		<Fragment>
+			<View style={styles.container}>
+				<View style={styles.header}>
+					<TouchableOpacity onPress={navigation.goBack} style={styles.navbarRightButton}>
+						<MaterialIcon name="close" size={15} style={styles.closeIcon} />
+					</TouchableOpacity>
+					<Icon name="qrcode" size={42} style={styles.qrcode} color={colors.text.default} />
+				</View>
+				{accounts.length <= 0 ? (
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/index.tsx:263]: Looks like this is correct, the condition is to specifically to check if the length is greater than 0

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+					<Icon name={'chevron-left'} color={colors.primary.default} />
+					<Text style={styles.paginationText}>{strings('connect_qr_hardware.prev')}</Text>
+				</TouchableOpacity>
+				<TouchableOpacity style={styles.paginationItem} onPress={nextPage}>
+					<Text style={styles.paginationText}>{strings('connect_qr_hardware.next')}</Text>
+					<Icon name={'chevron-right'} color={colors.primary.default} />
+				</TouchableOpacity>
+			</View>
+
+			<View style={styles.bottom}>
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx:151]: [70604c3](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/70604c34edecd39f737a463d5c9b8bfc52de082e)

File: app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+	prevPage: () => void;
+	toggleAccount: (index: number) => void;
+	onUnlock: () => void;
+	onForget: () => void;
+}
+
+const createStyle = (colors: any) =>
+	StyleSheet.create({
+		container: {
+			width: '100%',
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/SelectQRAccounts/index.tsx:29]: This breaks the UI, noted to improve in the next release

File: app/components/Views/ConnectQRHardware/Instruction/index.tsx [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+			alignItems: 'center',
+		},
+		container: {
+			width: '100%',
+			flexDirection: 'column',
+			alignItems: 'center',
+			paddingHorizontal: 32,
+		},
+		scrollWrapper: {
+			width: '100%',
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/Instruction/index.tsx:33]: This breaks the UI, noted to improve in the next release

File: app/components/Views/ConnectQRHardware/Instruction/index.tsx [hunk 25]
```diff
@@ -0,0 +1,128 @@
...
+		},
+		button: {
+			backgroundColor: colors.primary.default,
+			width: '80%',
+			borderRadius: 25,
+			height: 50,
+			alignItems: 'center',
+			justifyContent: 'center',
+			padding: 12,
+			marginBottom: 32,
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Views/ConnectQRHardware/Instruction/index.tsx]: Noted for next release or spot check in release build

File: app/components/UI/QRHardware/AnimatedQRScanner.tsx [hunk 26]
```diff
@@ -0,0 +1,239 @@
...
+	},
+	hint: {
+		backgroundColor: colors.whitetransparent,
+		width: '100%',
+		height: 120,
+		alignItems: 'center',
+		justifyContent: 'center',
+	},
+	hintText: {
+		width: 240,
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/QRHardware/AnimatedQRScanner.tsx:62]: Noted for next release or spot check in release build

File: app/components/UI/AccountList/AccountElement/index.js [hunk 27]
```diff
@@ -161,7 +163,7 @@ class AccountElement extends PureComponent {
 								)}
 							</View>
 						</View>
-						{!!imported && <View style={styles.importedView}>{imported}</View>}
+						{!!tag && tag}
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/AccountList/AccountElement/index.js:166]: Noted for next release or spot check in release build

File: app/components/UI/TransactionElement/index.js [hunk 28]
```diff
@@ -269,12 +273,18 @@ class TransactionElement extends PureComponent {
 							</ListItem.Amounts>
 						)}
 					</ListItem.Content>
-					{!!renderTxActions && (
+					{!!renderNormalActions && (
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/TransactionElement/index.js]: [b166e89](https://github.com/MetaMask/metamask-mobile/pull/3915/commits/b166e8922b5e1453a92244c2fc465baf96e2e338)

File: app/components/UI/QRHardware/QRSigningDetails.tsx [hunk 29]
```diff
@@ -0,0 +1,272 @@
...
+			marginVertical: 24,
+			alignItems: 'center',
+			...fontStyles.normal,
+			fontSize: 14,
+		},
+		descriptionTighten: {
+			marginVertical: 12,
+			fontSize: 12,
+		},
+		padding: {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/QRHardware/QRSigningDetails.tsx:86]: Noted for next release or spot check in release build

File: app/components/UI/QRHardware/QRSigningDetails.tsx [hunk 30]
```diff
@@ -0,0 +1,272 @@
...
+					setErrorMessage(strings('transaction.no_camera_permission_android'));
+				}
+			});
+		}
+	}, [hasCameraPermission]);
+
+	const [hasSentOrCanceled, setSentOrCanceled] = useState(false);
+
+	useEffect(() => {
+		navigation.addListener('beforeRemove', (e) => {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/QRHardware/QRSigningDetails.tsx:171]: Noted for next release or spot check in release build

File: app/components/UI/QRHardware/QRSigningModal/index.tsx [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+
+const createStyles = (colors: any) =>
+	StyleSheet.create({
+		modal: {
+			margin: 0,
+			justifyContent: 'flex-end',
+		},
+		contentWrapper: {
+			justifyContent: 'flex-end',
+			height: 600,
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/UI/QRHardware/QRSigningModal/index.tsx:24]: Noted for next release or spot check in release build

File: app/components/Nav/Main/RootRPCMethodsUI.js [hunk 32]
```diff
@@ -744,4 +763,4 @@ const mapDispatchToProps = (dispatch) => ({
 	networkSwitched: ({ networkUrl, networkStatus }) => dispatch(networkSwitched({ networkUrl, networkStatus })),
 });
 
-export default connect(mapStateToProps, mapDispatchToProps)(RootRPCMethodsUI);
+export default connect(mapStateToProps, mapDispatchToProps)(withQRHardwareAwareness(RootRPCMethodsUI));
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/components/Nav/Main/RootRPCMethodsUI.js:766]: Noted for next release or spot check in release build

**Review by cortisiko — APPROVED** [review]
🌮 🌮 🌮 🌮 🌮 🌮 🌮 🌮 🌮 🌮

**Review by Cal-L — APPROVED** [review]
Lgtm!

**Pull request merged**